engine/player/player.cpp

```diff
@@ -3029,6 +3029,123 @@ void player_t::init_finished()
       }
     } );
   }
+
+  if ( !precombat_state_map.empty() )
+  {
+    sim->error( "Warning: The 'override.precombat_state' option may not be fully supported for all buffs and cooldowns and and may produce incorrect or misleading results." );
+    std::unordered_map<buff_t*, std::tuple<int, double, timespan_t>> precombat_buff_state;
+
+    for ( const auto& v : precombat_state_map )
+    {
+      auto splits = util::string_split<std::string>( v.first, "." );
+
+      if ( splits.size() < 2 )
+      {
+        sim->error( "Invalid 'override.precombat_state' option: '{}'", v.first );
+        continue;
+      }
+
+      auto type = splits[ 0 ];
+      auto name = splits[ 1 ];
+
+      if ( type == "buff" )
+      {
+        if ( splits.size() != 3 ){
+          sim->error( "Invalid 'override.precombat_state' buff expression: '{}'", v.first );
+          continue;
+        }
+
+        auto buff = buff_t::find( this, name );
+        if ( !buff )
+        {
+          sim->error( "No buff found for 'override.precombat_state' buff expression: '{}'", name );
+          continue;
+        }
+
+        bool new_buff_state = precombat_buff_state.find( buff ) == precombat_buff_state.end();
+
+        if ( new_buff_state )
+        {
+          // Initialize the values for a buff's stack, value, and duration
+          // parameters to the default parameter values of buff_t::execute.
+          precombat_buff_state[ buff ] = { -1, buff_t::DEFAULT_VALUE(), timespan_t::min() };
+        }
+
+        if ( splits[ 2 ] == "stack" )
+        {
+          std::get<0>( precombat_buff_state[ buff ] ) = util::to_int( v.second );
+          continue;
+        }
+        else if ( splits [ 2 ] == "value" )
+        {
+          std::get<1>( precombat_buff_state[ buff ] ) = util::to_double( v.second );
+          continue;
+        }
+        else if ( splits [ 2 ] == "remains" )
+        {
+          std::get<2>( precombat_buff_state[ buff ] ) = timespan_t::from_seconds( util::to_double( v.second ) );
+          continue;
+        }
+
+        // If no part of the newly created state was set, it needs to be
+        // removed to avoid triggering the buff with all default values.
+        if ( new_buff_state )
```

Comment:
Adding the buff state and removing it again seems like very ugly and error prone control flow.

I see that it isn't as easy to merge the 3 statements because they parse v.second into different parts of the buff state, but it should still if possible.
If you have a function
```
buff_state_t& get_buff_state(buff_t* buff)
{
  bool new_buff_state = precombat_buff_state.find( buff ) == precombat_buff_state.end();

  if ( new_buff_state )
  {
    // Initialize the values for a buff's stack, value, and duration
    // parameters to the default parameter values of buff_t::execute.
    precombat_buff_state[ buff ] = { -1, buff_t::DEFAULT_VALUE(), timespan_t::min() };
  }
return precombat_buff_state[ buff ];
```

You can then do:
```

        if ( splits[ 2 ] == "stack" )
        {
          std::get<0>( get_buff_state[ buff ] ) = util::to_int( v.second );
          continue;
        }
```

And you even wrap the 3 ifs into a function you can use simple return statements instead of continue.